exchanges/gateio/gateio_ws_futures.go

```diff
@@ -0,0 +1,788 @@
+package gateio
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/account"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/fill"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/kline"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/ticker"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/trade"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const (
+	futuresWebsocketBtcURL  = "wss://fx-ws.gateio.ws/v4/ws/btc"
+	futuresWebsocketUsdtURL = "wss://fx-ws.gateio.ws/v4/ws/usdt"
+
+	futuresPingChannel            = "futures.ping"
+	futuresTickersChannel         = "futures.tickers"
+	futuresTradesChannel          = "futures.trades"
+	futuresOrderbookChannel       = "futures.order_book"
+	futuresOrderbookTickerChannel = "futures.book_ticker"
+	futuresOrderbookUpdateChannel = "futures.order_book_update"
+	futuresCandlesticksChannel    = "futures.candlesticks"
+	futuresOrdersChannel          = "futures.orders"
+
+	//  authenticated channels
+	futuresUserTradesChannel        = "futures.usertrades"
+	futuresLiquidatesChannel        = "futures.liquidates"
+	futuresAutoDeleveragesChannel   = "futures.auto_deleverages"
+	futuresAutoPositionCloseChannel = "futures.position_closes"
+	futuresBalancesChannel          = "futures.balances"
+	futuresReduceRiskLimitsChannel  = "futures.reduce_risk_limits"
+	futuresPositionsChannel         = "futures.positions"
+	futuresAutoOrdersChannel        = "futures.autoorders"
+)
+
+var defaultFuturesSubscriptions = []string{
+	futuresTickersChannel,
+	futuresTradesChannel,
+	futuresOrderbookChannel,
+	futuresOrderbookUpdateChannel,
+	futuresCandlesticksChannel,
+}
+
+// responseStream a channel thought which the data coming from the two websocket connection will go through.
+var responseStream = make(chan stream.Response)
+
+var futuresAssetType = asset.Futures
```

Comment:
asset. Futures is just the default values. its purpose is to distinguish which WSConnect method is used because there are some 'futures channels' push data processing functions to process both asset.Futures and asset.DeliveryFutures push datas.